SEED-platform/seed
# Labels to views

**WARNING** This PR involves making irreversible changes to the DB. See comments below.

#### Any background context you want to provide?
Labels are currently attached to canonical Properties and TaxLots (the base Property and TaxLot models). Since canonical Properties and TaxLots would potentially be shifted around during the upcoming year-over-year changes, it was decided to move Label associations to PropertyViews and TaxLotViews respectively. 

#1893 

#### What's this PR do?
This PR moves Label associations from canonical Properties and TaxLots to PropertyViews and TaxLotViews respectively.  

Specifically, there's code in the migration that goes through all the existing -Views, and for each, the code captures all the corresponding Property's (or TaxLot's) Label associations and applies these associations to the -View itself. Then, the migration runs to delete the join tables `seed_property_labels` and `seed_taxlot_labels` as well as all the records within them.
**WARNING: This is irreversible, and a backup of your database should be made.**

Upon changing Label relationships from Property/TaxLot to the corresponding -Views, the following existing **features** were transitioned to use this new relationship.
- Property & TaxLot Inventory List page View, Create, Add, & Remove Labels
- Property & TaxLot Inventory Detail page View, Create, Add, & Remove Labels
- Label can be associated to PropertyView or TaxLotView due to DQ Rule during post-import DQ check
- Property & TaxLot merges/unmerges combine and distribute Labels
- Property & TaxLot Labels are exported in CSV & GeoJSON exports
- Organization checks between Labels and target -View

Additionally, the following were checked to be sure nothing needed to be changed:
- Property & TaxLot Inventory List and Map pages filtering by Label
- Property/TaxLot Pairing
- Matching (unused "match" and "unmatch" methods in data_importer/tasks.py)
- Labels are not associated on import (according to DQ Rules). This is by design and is only mentioned here to say that this remained the same.

#### How should this be manually tested?
- Using an existing database with some data, add and/or make note of Properties/TaxLots and their associated Labels.
- Run migrations.
    - Again... **WARNING:** This is irreversible, and a backup of your database should be made.
- Be sure that from the user's perspective, Labels are still associated to their Properties/TaxLots.
- Monkey test any features involving Labels. See the list of features in the above section.
